### ISSUE TITLE
search_open and percent-encoding issue

### ISSUE THREAD

**Issue by sprnza** [luakit/luakit#616]
Hi there!
Just found out, that search_open() from window.lua fails if a custom engine link contains percent-encoding characters. For instance if I try to find a wiki.archlinux.org search string I get `https://wiki.archlinux.org/index.php?search=test&title=Special%3ASearch&go=Go` (for keyword "test"). If I copy and paste it to rc.lua I get errors from string.format() function. Replacing %3A with `:` solves the issue.

**Comment by aidanholm** [comment]
This is a side effect of how search engines work: they're a format string for a `string.format()` call, which means that `%` has special meaning. I'll add a pre-processing step that escapes everything but the critical `%s`

**Comment by sprnza** [comment]
```
607         local hex_to_char = function(x)
608             return string.char(tonumber(x, 16))
609         end
610         
611         local unescape = function(url)
612             return url:gsub("%%(%x%x)", hex_to_char)
613         end
614         local e = unescape(search_engines[engine]) or "%s"
```
Solves the issue

**Comment by aidanholm** [comment]
That will work for hex escapes, but will break if the hex digits are missing or invalid for some reason. I'd rather just fix the formatting issue, like so:
```lua
local l, r = e:match("^(.-)%%s(.-)$")
e = lousy.util.lua_escape(l) .. "%s" .. lousy.util.lua_escape(r)
```

**Comment by aidanholm** [comment]
Actually, that's not going to work, because it doesn't allow inserting a literal '%s', and there needs to be some level of compatibility escaping... urgh

**Comment by McSinyx** [comment]
> it doesn't allow inserting a literal '%s'

I believe that this is fine as `%s` should never appear in the URI as specified in RFC 3986: following the `%` should always be a 2-digit hex number and [`%` itself must be escaped](https://en.wikipedia.org/wiki/Percent-encoding). If there is a second `%s`, that'll likely to be because of user's typo or the carelessness of the search engine dev, which is none of our problem.

But I'd like to comment about a81d1f9: I think `e:gsub("%s", luakit.uri_encode(terms), 1)` could be used instead for readability.

**Comment by aidanholm** [comment]
That's a good point. Either way, any change to this system is going to break compatibility, so the current method should be fine. I agree that your version is _much_ nicer; thanks!

I think the API break can be resolved by falling back to the old `string.format()` code if the search engine contains `%%`, with a one-time warning.

**Comment by aidanholm** [comment]
Fixed at df794e4; thanks for reporting, and let me know if there's any more problems :)
